src/nvim/linematch.c

```diff
@@ -0,0 +1,484 @@
+}
+
+
+/// return matching characters between "s1" and "s2"
+/// between string "s1" and "s2".
+/// Consider the case of two strings 'AAACCC' and 'CCCAAA', the
+/// return value from this function will be 3, either to match
+/// the 3 C's, or the 3 A's.
+///
+/// @param s1
+/// @param s2
+int matching_characters(const char *s1, const char *s2)
+{
+  size_t s1len = STRLEN(s1), s2len = STRLEN(s2);
+  int *matrix[2];
+  matrix[0] = xmalloc(sizeof(int) * (s2len+1));
+  matrix[1] = xmalloc(sizeof(int) * (s2len+1));
+  bool icur = 1;  // save space by storing only two rows for i axis
+  for (size_t i = 0; i <= s1len; i++) {
+    icur = !icur;
+    for (size_t j = 0; j <= s2len; j++) {
+      if (i == 0) {
+        matrix[icur][j] = 0;
+      } else if (j == 0) {
+        matrix[icur][j] = 0;
+      } else {
+        matrix[icur][j] = 0;
+        // skip char in s1
+        if (matrix[!icur][j] > matrix[icur][j]) {
+          matrix[icur][j] = matrix[!icur][j];
+        }
+        // skip char in s2
+        if (matrix[icur][j-1] > matrix[icur][j]) {
+          matrix[icur][j] = matrix[icur][j-1];
+        }
+        // compare char in s1 and s2
+        if ( (s1[i-1] == s2[j-1])
+            && (matrix[!icur][j-1] + 1) > matrix[icur][j] ) {
+          matrix[icur][j] = matrix[!icur][j-1] + 1;
+        }
+      }
+    }
+  }
+  int rvalue = matrix[icur][s2len];
+  xfree(matrix[0]), xfree(matrix[1]);
+  return rvalue;
+}
+
+
+/// count the matching characters between a variable number of strings "stringps"
+/// mark the strings that have already been compared to extract them later
+/// without re-running the character match counting.
+/// @param stringps
+/// @param fromValues
+/// @param n
+/// @param comparison_mem
+int count_n_matched_chars(const char **stringps, const int *fromValues,
+                           const int n, int ***comparison_mem)
+{
+  int matched_chars = 0, pointerindex = 0, matched = 0;
+  for (int i = 0; i < n; i++) {
+    for (int j = i + 1; j < n; j++) {
+      if ( stringps[i] != NULL && stringps[j] != NULL ) {
+        int i1 = fromValues[i];  // index of where to get the buffer
+        int j1 = fromValues[j];
+        if (comparison_mem[pointerindex][i1][j1] == -1) {
+          comparison_mem[pointerindex][i1][j1] =
+            count_matched_chars(stringps[i], stringps[j]);
+        }
+        matched++;
+        matched_chars += comparison_mem[pointerindex][i1][j1];
+      }
+      pointerindex++;
+    }
+  }
+
+  // prioritize a match of 3 (or more lines) equally to a match of 2 lines
+  if (matched >= 2) {
+    matched_chars *= 2;
+    matched_chars /= matched;
+  }
+
+  return matched_chars;
+}
+
+/// allocate the memory for comparisons run with the diff linematch algorithm.
+/// this memory is used to prevent counting the matching characters on the same
+/// line twice
+/// @param diff_length
+/// @param nDiffs
+int ***allocate_comparison_mem(const int *diff_length, const int nDiffs)
```

Comment:
"mem" is too generic, all allocations are memory. This seems to allocate a set of buffers, so maybe `alloc_comparison_buffers` ?